onnxruntime/test/optimizer/nchwc_optimizer_test.cc

```diff
@@ -0,0 +1,839 @@
+    EXPECT_EQ(op_to_count["nchwc.Conv"], 4);
+    EXPECT_EQ(op_to_count["nchwc.ReorderInput"], 2);
+    EXPECT_EQ(op_to_count["nchwc.ReorderOutput"], 4);
+
+    // Verify that the weights and biases were converted once and reused.
+    std::unordered_set<const NodeArg*> weight_args;
+    std::unordered_set<const NodeArg*> bias_args;
+    const auto& graph = session.GetGraph();
+    for (auto& node : graph.Nodes()) {
+      if (node.Domain() == kMSNchwcDomain && node.OpType() == "Conv") {
+        EXPECT_EQ(node.InputDefs().size(), 3);
+        weight_args.emplace(node.InputDefs()[1]);
+        bias_args.emplace(node.InputDefs()[2]);
+      }
+    }
+    EXPECT_EQ(weight_args.size(), 1);
+    EXPECT_EQ(bias_args.size(), 1);
+  };
+
+  // Verify that a single weight tensor is reordered once.
+  NchwcOptimizerTester(build_test_case, check_nchwc_graph);
+}
+
+TEST(NchwcOptimizerTests, ShapeInferencing) {
+  auto build_test_case = [&](NchwcTestHelper& helper) {
+    ONNX_NAMESPACE::TypeProto type_proto;
+    type_proto.mutable_tensor_type()->set_elem_type(ONNX_NAMESPACE::TensorProto_DataType_FLOAT);
+    type_proto.mutable_tensor_type()->mutable_shape()->add_dim()->set_dim_value(1);
+    type_proto.mutable_tensor_type()->mutable_shape()->add_dim()->set_dim_value(32);
+    type_proto.mutable_tensor_type()->mutable_shape()->add_dim()->set_dim_param("input_height");
+    type_proto.mutable_tensor_type()->mutable_shape()->add_dim()->set_dim_param("input_width");
+
+    auto* input_arg = helper.MakeInput({1, 3, 50, 100}, type_proto);
+    auto* output_arg = helper.MakeOutput();
+
+    auto* conv1_output_arg = helper.MakeIntermediate();
+    auto& conv1_node = helper.AddConvNode(input_arg, conv1_output_arg, {48, 3, 3, 3});
+    conv1_node.AddAttribute("pads", std::vector<int64_t>{1, 1, 1, 1});
+
+    auto* conv2_output_arg = helper.MakeIntermediate();
+    auto& conv2_node = helper.AddConvNode(conv1_output_arg, conv2_output_arg, {64, 48, 3, 3});
+    conv2_node.AddAttribute("pads", std::vector<int64_t>{1, 1, 1, 1});
+
+    auto* conv3_output_arg = helper.MakeIntermediate();
+    auto& conv3_node = helper.AddConvNode(conv1_output_arg, conv3_output_arg, {64, 48, 3, 3});
+    conv3_node.AddAttribute("auto_pad", "SAME_UPPER");
+
+    auto* pool1_output_arg = helper.MakeIntermediate();
+    auto& pool1_node = helper.AddNode("MaxPool", {conv2_output_arg}, {pool1_output_arg});
+    pool1_node.AddAttribute("kernel_shape", std::vector<int64_t>{3, 3});
+    pool1_node.AddAttribute("pads", std::vector<int64_t>{1, 1, 1, 1});
+
+    auto* pool2_output_arg = helper.MakeIntermediate();
+    auto& pool2_node = helper.AddNode("MaxPool", {conv3_output_arg}, {pool2_output_arg});
+    pool2_node.AddAttribute("kernel_shape", std::vector<int64_t>{3, 3});
+    pool2_node.AddAttribute("auto_pad", "SAME_LOWER");
+
+    helper.AddNode("Add", {pool1_output_arg, pool2_output_arg}, {output_arg});
+  };
+
+  auto check_nchwc_graph = [&](NchwcInferenceSession& session) {
```

Comment:
What aspect of shape inferencing is this testing given the validation is only checking op counts?

Reply:
Added more comments and enhanced the test to show more clearly the benefit of doing this extra shape inferencing.